src/zui/components/ZUIVerticalBarChart/ZUIVerticalBarChartBars.tsx

```diff
@@ -0,0 +1,208 @@
+  const theme = useTheme();
+  const style = {
+    bar: {
+      backgroundColor: 'var(--barColor)',
+      borderRadius: 1,
+      flex: '1 0 auto',
+      marginInline: `min(${theme.spacing(0.5)}, 10%)`,
+      transition: 'background-color 0.3s',
+      width: 'auto',
+    },
+    barContainer: {
+      '&:hover': {
+        '--barColor': theme.palette.data.main,
+      },
+      '--barColor': theme.palette.data.mid1,
+      '@container (max-width: 4em)': {
+        '& .label': { opacity: 0 },
+      },
+      alignItems: 'end',
+      containerType: 'inline-size',
+      display: 'flex',
+      flex: 1,
+      height: '100%',
+      padding: 0,
+      position: 'relative',
+    },
+    bars: {
+      '&:has(li:hover)': {
+        '.label': {
+          opacity: 0,
+        },
+        'li:not(:hover)': {
+          '--barColor': theme.palette.data.final,
+        },
+      },
+      display: 'flex',
+      flex: '1',
+      height: visualizationHeight,
+      justifyContent: 'stretch',
+      listStyle: 'none',
+      margin: 0,
+      padding: 0,
+    },
+    hoveringElem: {
+      color: 'data.main',
+      paddingBlock: 0.25,
+      pointerEvents: 'none',
+      position: 'absolute',
+      transitionDuration: '.3s',
+      transitionProperty: 'opacity',
+      whiteSpace: 'nowrap',
+      width: 'auto',
+    },
+    hoveringLabel: {
+      top: '100%',
+    },
+    hoveringValue: {
+      top: 0,
+      translate: `0 -100%`,
+    },
+    label: {
+      '--translateX': '-50%',
+      '--translateY': '0',
+      color: 'text.secondary',
+      left: '50%',
+      maxWidth: '100%',
+      overflow: 'hidden',
+      paddingBlock: 0.25,
+      paddingInline: 0.5,
+      pointerEvents: 'none',
+      position: 'absolute',
+      textAlign: 'center',
+      textOverflow: 'ellipsis',
+      top: '100%',
+      transitionDuration: '0.3s',
+      transitionProperty: 'opacity',
+      translate: 'var(--translateX) var(--translateY)',
+      whiteSpace: 'nowrap',
+      width: 'auto',
+    },
+    root: {
+      position: 'relative',
+      width: '100%',
+    },
+  };
+
+  if (barIndex >= 0) {
+    const hoveredBar = containerRef.current?.querySelector(
+      `li:nth-of-type(${barIndex + 1})`
+    ) as HTMLDivElement;
+    const containerWidth = containerRef.current?.clientWidth || 0;
+    const barOffset = hoveredBar?.offsetLeft || 0;
+    const barWidth = hoveredBar?.clientWidth || 0;
+    const valueWidth = valueElem?.clientWidth || 0;
+    const labelWidth = labelElem?.clientWidth || 0;
+
+    const calculatedValuePosX = barOffset + barWidth / 2 - valueWidth / 2;
+    const boundedValuePosX = Math.max(
+      0,
+      Math.min(calculatedValuePosX, containerWidth - valueWidth)
+    );
+
+    const calculatedLabelPosX = barOffset + barWidth / 2 - labelWidth / 2;
+    const boundedLabelPosX = Math.max(
+      0,
+      Math.min(calculatedLabelPosX, containerWidth - labelWidth)
+    );
```

Comment:
Love that you are saving all these in clearly named variables! (or constants, to be more exact)